### ISSUE TITLE
Out of memory with large attachment

### ISSUE THREAD

**Issue by J-Jamet** [Kunzisoft/KeePassDX#115]
Create a better parser for elements in the database.

```
05-14 14:09:43.159 26057-26132/com.kunzisoft.keepass.libre E/art: Throwing OutOfMemoryError "Failed to allocate a 209715213 byte allocation with 4194304 free bytes and 36MB until OOM"
05-14 14:09:43.164 26057-26132/com.kunzisoft.keepass.libre E/com.kunzisoft.keepass.database.action.LoadDBRunnable: Votre appareil ne dispose pas de suffisamment de mémoire pour ouvrir votre base de données. Elle est probablement trop grosse pour votre appareil.
    java.lang.OutOfMemoryError: Failed to allocate a 209715213 byte allocation with 4194304 free bytes and 36MB until OOM
        at com.kunzisoft.keepass.stream.LEDataInputStream.readBytes(LEDataInputStream.java:109)
        at com.kunzisoft.keepass.database.load.ImporterV4.ReadInnerHeader(ImporterV4.java:232)
        at com.kunzisoft.keepass.database.load.ImporterV4.LoadInnerHeader(ImporterV4.java:219)
        at com.kunzisoft.keepass.database.load.ImporterV4.openDatabase(ImporterV4.java:190)
        at com.kunzisoft.keepass.database.load.ImporterV4.openDatabase(ImporterV4.java:77)
        at com.kunzisoft.keepass.database.Database.loadData(Database.java:173)
        at com.kunzisoft.keepass.database.Database.passUrisAsInputStreams(Database.java:152)
        at com.kunzisoft.keepass.database.Database.loadData(Database.java:119)
        at com.kunzisoft.keepass.database.Database.loadData(Database.java:107)
        at com.kunzisoft.keepass.database.action.LoadDBRunnable.run(LoadDBRunnable.java:71)
        at java.lang.Thread.run(Thread.java:818)
```

Linked to https://github.com/bpellin/keepassdroid/issues/287
#48

**Comment by J-Jamet** [comment]
The solution for this problem is to use an XML parser SAX and not DOM, but we will lose the speed of opening. It may be necessary to leave the choice to the user.

**Comment by theryecatcher** [comment]
I can confirm that I am facing the same issue. Database size is approximately 8 Mb with only dew jpgs for attachments on some entries and no additional file types. It opens well in Keepass2Andorid. It was also functioning perfectly alright till yesterday on KeepassDx. I am not sure what changed. I am definite that there were no additional entries in that time frame though.

**Comment by jimbt** [comment]
I'm a new user for KeePassDX.  I am seeing similar issues.  They don't seem to be primarily related to encryption, but compression.  With compression, I was having a hard time setting the Argon2 rounds, memory, and parallelism (or even AES-KDF rounds) low enough to allow the database to open on android.  If it did open, any editing would typically cause it to fail the next time.  However, at no time was KeePassXC on Linux unable to open the database.

With compression disabled, my database file is 440kB instead of 60kB, but I no longer have issues with KeePassDX (or KeePassDroid) on Android.  I was recently prompted to update the database encryption, but didn't realize it would cause problems until I made an entry edit.

**Comment by sevkin** [comment]
i can`t open 1,1Mb database on device with 2Gb :(

**Comment by J-Jamet** [comment]
I solved an OOM problem on the branch develop (https://github.com/Kunzisoft/KeePassDX/issues/256), I don't know if it's the same problem as this ticket, so if someone can compile the version of develop and confirm the resolution or no it would be nice. (or wait for version 2.5.0.0beta23)

**Comment by J-Jamet** [comment]
Normally corrected, tell me if it's ok with version 2.5.0.0beta23
